Added support for single_border_width and single_margin for VerticalTile layout

matthewsia98 opened the pull request:
Copy of previous PR because I had previously pushed to my fork's master branch.

Added single_border_width and single_margin keys for VerticalTile layout.

elParaguayo commented:
I would still like a test added for this (in `test/layouts/test_verticaltile.py`. If you're not comfortable doing that then we can add it.

matthewsia98 commented:
> I would still like a test added for this (in `test/layouts/test_verticaltile.py`. If you're not comfortable doing that then we can add it.

I have added the tests for the new features.
I also went ahead and fixed the other tests. They were failing because they did not account for the default ```border_width = 1```.

elParaguayo commented on test/layouts/test_verticaltile.py:
```diff
@@ -41,20 +41,31 @@ class VerticalTileConfig(Config):
 def test_verticaltile_simple(manager):
     manager.test_window("one")
-    assert_dimensions(manager, 0, 0, 800, 600)
+    assert_dimensions(manager, 0, 0, 798, 598)
```
Why has this test changed?

elParaguayo commented:
> > I would still like a test added for this (in `test/layouts/test_verticaltile.py`. If you're not comfortable doing that then we can add it.
> 
> I have added the tests for the new features. I also went ahead and fixed the other tests. They were failing because they did not account for the default `border_width = 1`.

They weren't failing before...

matthewsia98 commented:
> > > I would still like a test added for this (in `test/layouts/test_verticaltile.py`. If you're not comfortable doing that then we can add it.
> > 
> > 
> > I have added the tests for the new features. I also went ahead and fixed the other tests. They were failing because they did not account for the default `border_width = 1`.
> 
> They weren't failing before...

Ah I know what's going on. 
Before adding the ```single_border_width``` option, it would effectively behave like ```single_border_width = 0```.

Here is the relevant code on master in verticaltile.py
```            
# width
if n > 1:
    width = screen_rect.width - self.border_width * 2
else:
    width = screen_rect.width
```

But in contrast, if we take a look at the MonadTall layout, we can see that if ```single_border_width``` is not specified, it will effectively behave like ```single_border_width = border_width```.

<img width="404" alt="image" src="https://user-images.githubusercontent.com/71731492/201709320-64b944cf-0487-480d-8331-096d63eac639.png">

So I guess there needs to be a decision made on how VerticalTile will behave if ```single_border_width``` is not specified.

elParaguayo commented:
There should be no change for users i.e. without setting the parameters that you've added, the behaviour should be identical to the current behaviour unless it's clearly wrong.

matthewsia98 commented:
Alright I have reverted the changes to the original tests and made the necessary fixes to keep old behavior.

I would like to just point out the inconsistency in behavior between ```VerticalTile``` and ```MonadTall``` again.

For ```MonadTall```, if we don't specify the ```single_border_width``` or ```single_margin``` keys, those keys will get their default values of ```None``` replaced with the values of ```border_width``` and ```margin``` respectively.

For ```VerticalTile```, if we don't specify the ```single_border_width``` or ```single_margin``` keys, those keys will keep their default value of ```None``` effectively setting them to 0.


As for which one is "correct" behavior, that's not for me to decide.

elParaguayo commented:
Thanks. Inconsistency between different layouts is not ideal and should be something we look at. However, consistency in the behaviour of a layout when adding new features is important too.

Have restarted the tests so let's see how they get on.

elParaguayo commented:
You need to run `black` on your changed files as there are a couple of formatting aspects causing our checks to fail.

matthewsia98 commented on test/layouts/test_verticaltile.py:
```diff
@@ -90,3 +101,23 @@ def test_verticaltile_window_focus_cycle(manager):
+@verticaltile_single_border_config
+def test_verticaltile_single_border(manager):
+    manager.test_window("one")
+    assert_dimensions(manager, 0, 0, 796, 596)
```
Here's how it renders on my end.

Single Border
<img width="960" alt="image" src="https://user-images.githubusercontent.com/71731492/201948796-8e03ca27-f73f-4c12-b2c8-fce170492eff.png">

Normal Border
<img width="960" alt="image" src="https://user-images.githubusercontent.com/71731492/201948919-747a83f8-62cd-4d1a-8446-d1ebe07a44e6.png">

matthewsia98 replied:
Not sure what the code does exactly, but judgying by the presence of `outer_w` and `outer_h`, it does seem like borders are painted outside the actual window

<img width="619" alt="image" src="https://user-images.githubusercontent.com/71731492/202785351-164effb2-6d30-4060-b8c7-803e339381a2.png">

m-col replied:
Yes exactly right, the width and height are 'internal' dimensions that demarcate the contents of the window i.e. where the client renders. That would then be what `info()` reports in the tests too. It is a bit weird I agree, however these are the values we're sending in configure requests to the X server, meaning they are essentially X protocol abstractions. Of course that doesn't mean our `info()` command needs to align with that.

m-col reviewed (changes requested):
> For `MonadTall`, if we don't specify the `single_border_width` or `single_margin` keys, those keys will get their default values of `None` replaced with the values of `border_width` and `margin` respectively.
> 
> For `VerticalTile`, if we don't specify the `single_border_width` or `single_margin` keys, those keys will keep their default value of `None` effectively setting them to 0.

As the latter are new additions, I think it would be preferable to implement them in line with the existing options of `MonadTall`. It's also consistent with some other options e.g. `margin_on_single` in `Columns`, which if left as its default value of `None` is replaced by the less specific `margin` option's value.

m-col reviewed (approved):
All LGTM thanks! I've enabled auto-merge so it will merge if CI passes.
